### DISCUSSION TITLE
http accessing file from a bucket

### DISCUSSION THREAD

**Discussion by AllanJard** [seaweedfs/seaweedfs#4004] [Q&A] [answered]
I know this is going to be really basic, and I'm missing something, but I'm getting totally confused with accessing a file I've successfully uploaded (thank you!) to a seaweedfs S3 bucket.

The response from the S3 upload API is:

```
{
	"$metadata": {
		"httpStatusCode": 200,
		"requestId": "1669119158228673318",
		"attempts": 1,
		"totalRetryDelay": 0
	},
	"ETag": "\"e1bfa4f75c27f1ee440e6a5ab6c226b0\"",
	"Bucket": "ce301f0c-6a5e-11ed-b9f4-ab7e496be0b3",
	"Key": "65795350-86bf-43a1-98c4-2dd2249ea10c",
	"Location": "http://127.0.0.1/ce301f0c-6a5e-11ed-b9f4-ab7e496be0b3/65795350-86bf-43a1-98c4-2dd2249ea10c"
}
```

So I assumed that I'd need to add the volume server or filer server's port number:

```
Volume:
400 Bad request - http://127.0.0.1:8080/ce301f0c-6a5e-11ed-b9f4-ab7e496be0b3/65795350-86bf-43a1-98c4-2dd2249ea10c
400 Bad request - http://127.0.0.1:8080/buckets/ce301f0c-6a5e-11ed-b9f4-ab7e496be0b3/65795350-86bf-43a1-98c4-2dd2249ea10c

Filer:
404 Not Found   - http://127.0.0.1:8888/ce301f0c-6a5e-11ed-b9f4-ab7e496be0b3/65795350-86bf-43a1-98c4-2dd2249ea10c
200 OK          - http://127.0.0.1:8888/buckets/ce301f0c-6a5e-11ed-b9f4-ab7e496be0b3/65795350-86bf-43a1-98c4-2dd2249ea10c
404 Not Found   - http://127.0.0.1:8888/buckets/ce301f0c-6a5e-11ed-b9f4-ab7e496be0b3/65795350-86bf-43a1-98c4-2dd2249ea10c/test.png
```

I'd been hoping to do something like [in the readme](https://github.com/seaweedfs/seaweedfs#read-file) adding the original file name to the url. But that appears to be through the volume server, which I can't seem to find the right URL for.

I suppose moreover, even if it is possible through the volume server, should I be serving data publicly via filer or the volume server?

**Comment by jerebear12** [comment, marked as answer]
You can only read from the volume server using the [*internal file id*](https://github.com/seaweedfs/seaweedfs/wiki/Volume-Server-API#volume-server-reads), not the filename as described by you. The Filer allows you to request files by the filename, as provided by you. This is why you are unable to get the image from the volume using the API but the filer works just fine.

Also, request 4 works over request 3 because when using the S3 API you are putting the file to a bucket no matter what. When requesting the file you need to prefix the path with the buckets/ folder to retrieve images in that bucket. Request 3 is asking for the file in the base folder on the Filer, which contains the folder buckets/ not the bucket folder itself. 

Yes, it makes more sense to serve files from just the Filer or the Filer using the S3 gateway.

Let me know if you have any more questions.

**Comment by AllanJard** [comment]
That's awesome - thanks so much for the clarification.
